### PR TITLE
[Android] Fix #288 Reset to default layer when switching input field contexts

### LINKED ISSUE #288: Keyboard should reset to default layer when new input field focused
<a href="https://github.com/andjc"></img></a> **Issue by [andjc](https://github.com/andjc)**
_Tuesday Jun 10, 2014 at 21:42 GMT_
_Originally opened as KD-29_

----
also one comment on differences between KMT and most other android keyboard layouts
KMT retains the last layer used so if the last thing i typed on eur keyboard is a digit, next time i go to type it starts on the digit layer

most other keyboard/input systems default to normal layer, this is most noticeable when the last character you typed was on the digit layer on login, and once you go into an app later and start typing, it starts on that digit layer

SK adds:
> KMEA should be able to tell KMW Engine to reset to default layer.

### PR DESCRIPTION
**Pull request by darcywong00**
Add resetContext() to the keyman engine API so embedded platforms can reset to the default layer when user changes contexts.

### CODE REVIEW

File: android/KMEA/app/src/main/assets/keyboard.html [hunk 1]
```diff
@@ -90,6 +90,11 @@
       window.jsInterface.insertText(dn, s);
     }
 
+    function resetContext() {
+      var kmw = window['tavultesoft']['keymanweb'];
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/assets/keyboard.html]: We don't need to use this model of referencing within **keyboard.html**. The `['name']` approach is the hoopy way that Google Closures identifies names that should not be minified. **keyboard.html** will never be minified, so we can instead say `tavultesoft.keymanweb.resetContext();` and avoid `['']`land altogether here. Appreciate there may be other areas in keyboard.html that already do this, but let's not perpetuate it.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 2]
```diff
@@ -1730,6 +1730,18 @@ public static boolean updateSelectionRange(KeyboardType kbType, int selStart, in
     return result;
   }
 
+  public static void resetContext(KeyboardType kbType) {
+    if (kbType == KeyboardType.KEYBOARD_TYPE_INAPP) {
+      if (InAppKeyboard != null && InAppKeyboardLoaded && !InAppKeyboardShouldIgnoreTextChange) {
+        InAppKeyboard.loadUrl("javacsript:resetContext()");
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: typo: `javacsript:` should be `javascript:`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 3]
```diff
@@ -1730,6 +1730,18 @@ public static boolean updateSelectionRange(KeyboardType kbType, int selStart, in
     return result;
   }
 
+  public static void resetContext(KeyboardType kbType) {
+    if (kbType == KeyboardType.KEYBOARD_TYPE_INAPP) {
+      if (InAppKeyboard != null && InAppKeyboardLoaded && !InAppKeyboardShouldIgnoreTextChange) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: Why is this testing `InAppKeyboardShouldIgnoreTextChange` but the other one tests `SystemKeyboardShouldIgnoreSelectionChange`? Do those two variables serve the same purpose?

File: web/source/keymanweb.js [hunk 4]
```diff
@@ -1318,7 +1317,26 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
...
   
     // End of I3363 (Build 301) additions
-    
+
+    /**
+     * Function     resetContext
+     * Scope        Public
+     * Description  Revert OSK to default layer and clear any deadkeys and modifiers
+     */
+    keymanweb['resetContext'] = keymanweb.resetContext = function()
```
Comment (mcdurdin) [inline web/source/keymanweb.js]: @jahorton Can you review this function for correctness? I am a little unsure of the `osk._Visible = 0` line and also you will be more aware of any other state variables that may need reset.

File: web/source/kmwembedded.js [hunk 5]
```diff
@@ -32,39 +32,39 @@
     if(device.app.indexOf('Mobile') >= 0) device.formFactor='phone';
     if(device.app.indexOf('Tablet') >= 0) device.formFactor='tablet';
     device.browser='native';
-  }
+  };
```
Comment (mcdurdin) [inline web/source/kmwembedded.js]: These `;` additions are good tidyup but really belong in their own PR. Happy to leave them for now, but I think code cleanup like this should almost always be in its own PR in future.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 3 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: Oops, copy-paste error. I think I should use `SystemKeyboardShouldIgnoreTextChange`

**Comment by darcywong00** [comment]
@jahorton Can you also provide me with a test case scenario of checking that deadkeys get reset?

**Comment by darcywong00** [comment]
I'm pushing commit a18bb2a2fdd298d7997207d4cf201c7f989abc4d so people can test the in-app keyboard behavior when switching text areas.

We'll need to remember to revert this before merging the PR

**Comment by jahorton** [comment]
@darcywong00 Deadkey testing:

Pull up the "French-French" keyboard and go to the Alt-Ctrl layer.  Both the '~' key and '`' key are deadkeys and only appear in combination with input vowels typed afterward.

So, try '~' through that Alt-Ctrl layer, then go back to the default layer and enter 'a'.

File: web/source/keymanweb.js [hunk 6]
```diff
@@ -1318,7 +1317,25 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
...
+    {
+      osk.layerId = 'default';
+      osk.layerIndex = 0;
```
Comment (jahorton) [inline web/source/keymanweb.js]: No need to bother with the `osk.layerIndex = 0;` line - that's automatically managed by osk._Show() once `osk.layerId` is set.

Also, check the file formatting!  I believe our official policy says the opening brace should be at the end of the previous line.

File: web/source/keymanweb.js (hunk 4 above)
Comment (jahorton) [reply web/source/keymanweb.js]: Done.  It looks correct to the best of my understanding; it'll even work correctly with the upcoming chirality changes.  `_ResetVKShift` is where that magic is.

File: web/source/keymanweb.js (hunk 6 above)
Comment (darcywong00) [reply web/source/keymanweb.js]: hmm. So for this set up:

1. Alt-Ctrl layer
2. `~`
3. Switch text field

The keyboard is staying in the Alt-Ctrl layer.  I'll have to see what I'm missing...

**Comment by darcywong00** [comment]
iOS has the same issue, so I've created #306 for that platform.

**Pull request merged**